search: Deduplicate results w.r.t. repositories

keegancsmith opened the pull request:
During a rebalancing event repositories can be indexed on multiple nodes. This
is temporary, but will lead to us returning duplicate results. To avoid this
situation we do not aggregate a filematch from a shard if another shard has
already returned a result for the same repository.

Notes:
- RFC 30 encourages we split up the RepoSet to avoid duplicate results. This
  is something we may still do, but this is a more incremental change and will
  allow us to focus on making indexing more intelligent first.
- The map this allocates can be as large as our RepoSet. However, this should
  be much smaller in practice, and will always be smaller than the memory used
  by []FileMatch. As such this shouldn't impact performance.
- We don't dedup List since the only use of it already does
  deduplication. Avoiding this coupling we will address later and is noted in
  the code.
- The Stats struct has fields for MatchCount and FileCount. We don't update
  these fields w.r.t. deduplication since we do not consume them. Related to
  above we can address this at a later stage.

Test plan: Unit tests

Part of https://github.com/sourcegraph/sourcegraph/issues/5725

keegancsmith commented:
This PR originally had some optimizations to dedup. But I was interested if they are actually necessary. After benchmarking, it seems not. So I dropped the commit from the PR and force pushed.

tsenart commented on internal/search/backend/horizontal.go:
```diff
@@ -46,14 +46,17 @@ func (s *HorizontalSearcher) Search(ctx context.Context, q query.Q, opts *zoekt.
 		}(c)
 	}
 
+	// During rebalancing a repository can appear on more than on replica.
```
```suggestion
	// During rebalancing a repository can appear on more than one replica.
```

tsenart commented on internal/search/backend/horizontal.go:
```diff
@@ -186,3 +192,31 @@ func equalKeys(a map[string]zoekt.Searcher, b map[string]struct{}) bool {
+
+	// Update seenRepo now, so the next call of dedup will contain the
+	// repos.
+	for _, fm := range dedup {
```
Do we need to iterate over dedup separately? Would it be enough to add `seenRepo[fm.Repository] = struct{}{}` to line 213?

keegancsmith replied:
no, because then multiple file matches from the same repo would be removed (see test cases)

keegancsmith commented:
> This PR originally had some optimizations to dedup. But I was interested if they are actually necessary. After benchmarking, it seems not. So I dropped the commit from the PR and force pushed.

The benchmark was broken! In fact the perf commit did make quite a significant difference, so I've added it back :)

```
name     old time/op  new time/op  delta
Dedup-8   464µs ± 4%   310µs ± 4%  -33.09%  (p=0.008 n=5+5)
```
